streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/AssignmentInfo.java

```diff
@@ -130,8 +162,24 @@ public int latestSupportedVersion() {
      */
     public ByteBuffer encode() {
         final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try {
+            baos.write(toBytes(usedVersion));
+        } catch (final IOException ex) {
+            throw new TaskAssignmentException("Failed to encode AssignmentInfo", ex);
+        }
 
-        try (final DataOutputStream out = new DataOutputStream(baos)) {
+        final DataOutputStream out;
+        try {
+            if (usedVersion >= 5) {
+                out = new DataOutputStream(new GZIPOutputStream(baos));
```

Comment:
I am wondering if we can defer compression until later.. Did a small microbenchmark of the implementation here.. 
`numTasks : 100000 , numHosts: 1000 , AssignmentInfo Size in bytes : 97465`

![Screen Shot 2019-07-30 at 2 43 57 PM](https://user-images.githubusercontent.com/1179324/62167963-377b8500-b2d9-11e9-88e4-39dac28e39ae.png)

Let me know if this makes sense..

Reply:
I agree 97kb still looks ok, but when you look at the whole assignment object, which has assignmentInfo for each consumer. Can you check size of assignment object in StreamsPartitionAssignor.java, it will be almost 1000*assignmentinfo in above example. Then compression will make more sense as it might reduce assignment object from 97Mb to around 20-30Mb.

Reply:
Let me test that out as well and circle back.

Reply:
https://issues.apache.org/jira/browse/KAFKA-7149?focusedCommentId=16615271&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16615271  is what you are referring to I guess. 

you are right. even for something similar to the what's reported KAFKA-6976 (256*4 topic partitions across 400 hosts), we would exceed the ~1MB limit, while logging the assignment information to kafka

`numTasks : 1000 , numHosts: 400 , Single AssignmentInfo Size in bytes : 3296, Total AssignmentInfo size bytes : 1318400 `

IMO, the longer term solution is to change the way the `Map<String, Assignment>` is logged to kafka and this is a protocol change, which needs to be more broadly discussed. In the meantime, encoding taskIDs (this PR) and follow up with ways to turn on compression on internal topics, should achieve the intent of this JIRA and give us a lot of mileage ahead?

Reply:
I have some quick sync with @vinothchandar about this, and I think I agree that in the long run we should avoid sending the duplicated host-info over the wire, which would require us having a `global user data` field sent by the leader and propagated to all the members, but only written once over the wire and on the broker side.

Since this is requiring a protocol / broker side change, I think atm we can try to complete this PR for client-side only, while letting brokers to enable compression on internal topics probably, and then work on this bigger change sooner after this change.

Reply:
Yeah, this was one of the way we thought of reducing total assignment and since it required broker level changes, we decided to go ahead with the current approach. Long term as you said should be that as we are replicating same information for each consumer. But with the tests I did with current changes, our assignment object reduced from 196MB to 8MB, which was good enough change.